### ISSUE TITLE
Overflow error with new way of p-value calculation in kendall tau correlation for perfectly monotonic vectors.

### ISSUE THREAD

**Issue by christosbampis** [scipy/scipy#9611]
My issue is about overflow error caused by zero discordant pairs in scipy version 1.2.0, in the improved p-value calculation of the kendall's tau function. In the case of two perfectly monotonic vectors x, y, c becomes 0 and the if-else condition calculates a factorial. In version 1.1.0, this refined term calculation is not present and so this issue does not occur there. If others can reproduce this issue, it seems that the solution is simple: when c becomes 0 (when dis is 0), p-value converges to zero very rapidly for sizes > 15 etc., so an if-statement could take care of that. Another solution could be to flag large sample sizes with method='asymptotic' being used instead.

(The c = 1 case seems to be similar in my mind.)

```
elif c == 0:
    pvalue = 2.0/np.math.factorial(size)
elif c == 1:
    pvalue = 2.0/np.math.factorial(size-1)
```

### Reproducing code example:

Sample code to reproduce the problem
```
from scipy import stats
kendall, _ = stats.kendalltau(np.arange(200), np.arange(200))
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/scipy/stats/stats.py", line 3632, in kendalltau
    pvalue = 2.0/np.math.factorial(size)
OverflowError: long int too large to convert to float
```
Note that in this example, size is equal to 200.

### Scipy/Numpy/Python version information:
```
>>> scipy.__version__
'1.2.0'
```

This error does not occur in 1.1.0.

**Comment by chrisb83** [comment]
Thanks for the bug report. `np.longdouble(np.math.factorial(170))` works, but for 171 it also overflows.

Current condition to determine the method is as follows:

```
if method == 'auto':
        if (xtie == 0 and ytie == 0) and (size <= 33 or min(dis, tot-dis) <= 1):
            method = 'exact'
```

**Comment by bashtage** [comment]
Should use exp(gammaln(size)) which will give +inf.

**Comment by vigna** [comment]
Please fix this. It's embarrasing.

**Comment by NathanRxl** [comment]
I just bumped into a variant of this issue in 1.5.2.

When using stats.kendalltau, the same error `OverflowError: long int too large to convert to float` is raised (for the same reasons as the ones described in this issue) when passing these conditions https://github.com/scipy/scipy/blob/master/scipy/stats/stats.py#L4450-L4456

contains_nan == True
nan_policy == 'omit' (not the default)
variant == 'b' (default)

The problem is that the same code pattern that led to the issue reported here is present in https://github.com/scipy/scipy/blob/master/scipy/stats/mstats_basic.py#L640 . This one performs the factorial without handling the case where n is too large.

Same fix as proposed in #10056 (but on mstats_basic) should fix the remaining variants of this issue
